SAML2/XML/Signature.hs

```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
 -- | take a public key and an xml node ID that points to the sub-tree that needs to be signed, and
 -- return @Right ()@ if it is signed with that key.  otherwise, return a (hopefully helpful) error.
 -- use this if you want to verify signatures, and ignore the rest of this module if you can.
+--
+-- how does this work?:
+--   * dig for the subtree of the input with an ID attribute containing xid (the "signed subtree")
+--   * parse the 'Sigature' subtree in that subtree (we only do envelopped signatures)
+--   * get the canonicalized 'SignedInfo' subtree of the signed subtree as bytestring.
+--   * call 'verifyReference' on all 'Reference's contained in the parsed signature to make sure input is intact.
+--   * call 'verifyBytes' on the canonicalized 'SignedInfo' to make sure the signature is valid.
+--
+-- the canonicalizations given in the signature are applied to the signed info; the transforms
+-- are applied to the signed subtrees.  (this is confusing because one of the transforms is
+-- usually a form of canonicalization, but it makes sense if you accept the premise that any
+-- of this does.)
 verifySignature :: PublicKeys -> String -> HXT.XmlTree -> IO (Either SignatureError ())
 verifySignature pks xid doc = runExceptT $ do
-  x :: HXT.XmlTree
-    <- case HXT.runLA (getID xid) doc of
-      [x] -> return x
-      _ -> throwError SignedElementNotFound
-  sx :: HXT.XmlTree
-    <- case child "Signature" x of
-      [sx] -> return sx
-      _ -> throwError SignatureNotFoundOrEmpty
-  s@Signature{ signatureSignedInfo = si } :: Signature
-    <- case docToSAML sx of
-      Left err -> throwError . SignatureParseError $ show err
-      Right v -> pure v
-  six :: BS.ByteString
-    <- failWith SignatureCanonicalizationError
-      $ (applyCanonicalization (signedInfoCanonicalizationMethod si) (Just xpath) $ DOM.mkRoot [] [x])
-  rl :: NonEmpty.NonEmpty (Either String String)
-    <- failWith (SignatureVerifyReferenceError . (show (signedInfoReference si) <>))
-      $ mapM (`verifyReference` x) (signedInfoReference si)
-
-  when (null rl) $
-    throwError . SignatureVerifyNoReferences $ show rl
-  unless (all isRight rl) $
-    throwError . SignatureVerifyBadReferences $ show (signedInfoReference si, rl)
-  unless (elem (Right xid) rl) $
-    throwError . SignatureVerifyInputNotReferenced $ show rl
+  signedSubtree :: HXT.XmlTree
+    <- do
+        mdoc' <- liftIO . try $ fixNamespaces doc
+        doc' <- case mdoc' of
+          Right x
+            -> pure x
+          Left (err :: SomeException)
```

Comment:
It'd be nice to move the `SomeException` up to `try @SomeException` so the exception type stays close to the `try` and makes it clear that we plan to catch ALL exceptions 😄